crypto/asn1/a_time.c

```diff
@@ -509,14 +511,60 @@ int ossl_asn1_time_print_ex(BIO *bp, const ASN1_TIME *tm)
         }
 
         return BIO_printf(bp, "%s %2d %02d:%02d:%02d%.*s %d%s",
-                          _asn1_mon[stm.tm_mon], stm.tm_mday, stm.tm_hour,
-                          stm.tm_min, stm.tm_sec, f_len, f, stm.tm_year + 1900,
-                          (gmt ? " GMT" : "")) > 0;
+                    _asn1_mon[stm.tm_mon], stm.tm_mday, stm.tm_hour,
+                    stm.tm_min, stm.tm_sec, f_len, f, stm.tm_year + 1900,
+                    (gmt ? " GMT" : "")) > 0;
     } else {
         return BIO_printf(bp, "%s %2d %02d:%02d:%02d %d%s",
-                          _asn1_mon[stm.tm_mon], stm.tm_mday, stm.tm_hour,
-                          stm.tm_min, stm.tm_sec, stm.tm_year + 1900,
-                          (gmt ? " GMT" : "")) > 0;
+                    _asn1_mon[stm.tm_mon], stm.tm_mday, stm.tm_hour,
+                    stm.tm_min, stm.tm_sec, stm.tm_year + 1900,
+                    (gmt ? " GMT" : "")) > 0;
+    }
+}
+
+/* returns 0 on BIO write error, else -1 in case of parse failure, else 1 */
+int ossl_asn1_time_print_ex_iso(BIO *bp, const ASN1_TIME *tm)
+{
+    char *v;
+    int gmt = 0, l;
+    struct tm stm;
+    const char upper_z = 0x5A, period = 0x2E;
+
+    /* ossl_asn1_time_to_tm will check the time type */
+    if (!ossl_asn1_time_to_tm(&stm, tm))
+        return BIO_write(bp, "Bad time value", 14) ? -1 : 0;
+
+    l = tm->length;
+    v = (char *)tm->data;
+    if (v[l - 1] == upper_z)
+        gmt = 1;
+
+    if (tm->type == V_ASN1_GENERALIZEDTIME) {
+        char *f = NULL;
+        int f_len = 0;
+
+        /*
+         * Try to parse fractional seconds. '14' is the place of
+         * 'fraction point' in a GeneralizedTime string.
+         */
+        if (tm->length > 15 && v[14] == period) {
+            f = &v[14];
+            f_len = 1;
+            while (14 + f_len < l && ossl_ascii_isdigit(f[f_len]))
+                ++f_len;
```

Comment:
If length is 17 then flen = 3.. Is this what you intended?

Reply:
Same comment as above